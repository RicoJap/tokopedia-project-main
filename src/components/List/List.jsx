import React from "react";
import MuiList from "@material-ui/core/List";
import MuiListItem from "@material-ui/core/ListItem";
import MuiListItemIcon from "@material-ui/core/ListItemIcon";
import MuiListItemText from "@material-ui/core/ListItemText";
import { makeStyles } from "@material-ui/core/styles";

import { isEmpty } from "../../utilities/check-if-empty.utils";
import { toTitleCase } from "../../utilities/string.utils";
import { PRIMARY_COLOR } from "../../utilities/theme.const";

const useListItemStyles = makeStyles({
  root: {
    backgroundColor: PRIMARY_COLOR,
    color: "#ffffff",
    opacity: 0.75,
  },
});

/**
 * List Component
 *
 * @since  15/11/2020
 * @access (public)
 *
 * @type     Functional Component
 * @param {Array}  listItems  Items to be rendered by the list
 * @param {NUMBER}  iconSelector  the key selector of each item object to get the icon for that particular item
 * @param {Function}  labelSelector  the key selector of each item object to get the label for that particular item
 *
 */

const List = ({
  listItems,
  iconSelector = "icon",
  labelSelector = "label",
}) => {
  const listItemStyles = useListItemStyles();

  return (
    <MuiList>
      {listItems.map((listItem, i) => {
        return (
          <MuiListItem key={i} classes={listItemStyles} divider>
            {!isEmpty(listItem[iconSelector]) && (
              <MuiListItemIcon>{listItem[iconSelector]}</MuiListItemIcon>
            )}
            <MuiListItemText data-testid={listItem[labelSelector]} primary={toTitleCase(listItem[labelSelector])} />
          </MuiListItem>
        );
      })}
    </MuiList>
  );
};

export default List;
