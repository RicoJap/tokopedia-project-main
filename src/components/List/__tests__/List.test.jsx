import "@testing-library/jest-dom";
import React from "react";
import renderer from "react-test-renderer";

import { fireEvent, render, screen, waitFor } from "@testing-library/react";

import List from "../List";

const mockListItems = [
  {
    name: "Mock List Item 1",
  },
  {
    name: "Mock List Item 2",
  },
  {
    name: "Mock List Item 3",
  },
];

const MockList = () => {
  return <List listItems={mockListItems} labelSelector="name" />;
};

describe("Snapshot testing", () => {
  const tree = renderer.create(<MockList />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("List Component renders correctly", () => {
  test("Check if List component is present in the document", () => {
    render(<MockList />);
    expect(screen.getByTestId("Mock List Item 1")).toBeInTheDocument();
    expect(screen.getByTestId("Mock List Item 2")).toBeInTheDocument();
    expect(screen.getByTestId("Mock List Item 3")).toBeInTheDocument();
  });
});
