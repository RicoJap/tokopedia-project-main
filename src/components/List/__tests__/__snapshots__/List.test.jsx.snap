// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<ul
  className="MuiList-root MuiList-padding"
>
  <li
    className="MuiListItem-root makeStyles-root-1 MuiListItem-gutters MuiListItem-divider"
    disabled={false}
  >
    <div
      className="MuiListItemText-root"
      data-testid="Mock List Item 1"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        Mock List Item 1
      </span>
    </div>
  </li>
  <li
    className="MuiListItem-root makeStyles-root-1 MuiListItem-gutters MuiListItem-divider"
    disabled={false}
  >
    <div
      className="MuiListItemText-root"
      data-testid="Mock List Item 2"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        Mock List Item 2
      </span>
    </div>
  </li>
  <li
    className="MuiListItem-root makeStyles-root-1 MuiListItem-gutters MuiListItem-divider"
    disabled={false}
  >
    <div
      className="MuiListItemText-root"
      data-testid="Mock List Item 3"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        Mock List Item 3
      </span>
    </div>
  </li>
</ul>
`;
